usl_pipeline/cloud_functions/main.py

```diff
@@ -903,3 +903,104 @@ def _write_as_npy(blob: storage.Blob, array: NDArray) -> None:
+        db, study_area_name, chunk_name
+    )
+    if chunk_metadata is None:
+        logging.info(
+            f"Chunk metadata is not registered for {study_area_name}/{chunk_name}"
+        )
+        return
+
+    if not chunk_metadata.needs_scaling:
+        logging.info(f"Chunk {study_area_name}/{chunk_name} doesn't need scaling")
+        return
+    study_area = metastore.StudyArea.get(db, study_area_name)
+
+    if blob_path.endswith(".npy"):
+        if study_area.chunk_x_count is None or study_area.chunk_y_count is None:
+            raise ValueError(
+                f"[Feature Rescaler] Study area {study_area_name} has no chunk counts"
+                + " in metadata"
+            )
+
+        feature_blobs = list(
+            feature_bucket.list_blobs(prefix=f"{study_area_name}/chunk_")
+        )
+
+        chunk_count = study_area.chunk_x_count * study_area.chunk_y_count
+        if chunk_count != len(feature_blobs):
+            raise ValueError(
+                f"[Feature Rescaler] Study area {study_area_name} has chunk counts"
+                + f" in metadata ({chunk_count}) different from number of stored"
+                + f" feature chunks ({len(feature_blobs)})"
+            )
+
+        # We're at the final step when all unscaled feature matrices are ready
+        for feature_blob in feature_blobs:
+            _, blob_chunk_name = _parse_chunk_path(feature_blob.name)
+            trigger_blob = feature_bucket.blob(
+                f"{study_area_name}/{blob_chunk_name}{_FEATURE_SCALING_TRIGGER_SUFFIX}"
+            )
+            trigger_blob.upload_from_string(data="")  # Empty file content
+    elif blob_path.endswith(_FEATURE_SCALING_TRIGGER_SUFFIX):
+        logging.info(
+            "[Feature Rescaler] Rescaling feature matrix %s/%s.npy",
+            study_area_name,
+            chunk_name,
+        )
+        feature_blob = feature_bucket.blob(f"{study_area_name}/{chunk_name}.npy")
+        output_blob = feature_bucket.blob(f"{study_area_name}/scaled_{chunk_name}.npy")
+        with feature_blob.open("rb") as in_fd, output_blob.open("wb") as out_fd:
+            feature_raster_transformers.rescale_feature_blob(in_fd, study_area, out_fd)
+        scaled_feature_matrix_path = (
+            f"gs://{feature_bucket.name}/{study_area_name}/scaled_{chunk_name}.npy"
+        )
+        metastore.StudyAreaSpatialChunk.update_scaling_done(
+            db, study_area_name, chunk_name, scaled_feature_matrix_path
+        )
+        # Deleting unscaled matrix
+        feature_blob.delete()
+        # Deleting trigger file
+        feature_bucket.blob(blob_path).delete()
+
+
+@functions_framework.cloud_event
+@_retry_and_report_errors()
+def rescale_feature_matrices(cloud_event: functions_framework.CloudEvent) -> None:
+    """Starts rescaling process once all unscaled feature matrix files are uploaded.
+
+    Output files with scaled feature matrices will be placed to the same folder where
```

Comment:
Let's describe the scale trigger files as well. I'd generally suggest being aggressive with the amount of comments you write in here.

Reply:
Done